src/awkward/_v2/operations/describe/ak_type.py

```diff
@@ -2,147 +2,154 @@
+
+    has type
+
+        3 * union[int64, string, var * int64]
+
+    but "union" is not a Datashape type-constructor. (Its syntax is
+    similar to existing type-constructors, so it's a plausible addition
+    to the language.)
+    """
+    if array is None:
+        return ak._v2.types.UnknownType()
+
+    elif isinstance(
+        array,
+        (
+            np.int8,
+            np.int16,
+            np.int32,
+            np.int64,
+            np.uint8,
+            np.uint16,
+            np.uint32,
+            np.uint64,
+            np.float32,
+            np.float64,
+            np.complex64,
+            np.complex128,
+            np.datetime64,
+            np.timedelta64,
+        ),
+    ):
+        return ak._v2.types.PrimitiveType(type.dtype2primitive[array.dtype.type])
+
+    elif isinstance(array, (bool, np.bool_)):
+        return ak._v2.types.PrimitiveType("bool")
+
+    elif isinstance(array, numbers.Integral):
+        return ak._v2.types.PrimitiveType("int64")
+
+    elif isinstance(array, numbers.Real):
+        return ak._v2.types.PrimitiveType("float64")
+
+    elif isinstance(
+        array,
+        (
+            ak._v2.highlevel.Array,
+            ak._v2.highlevel.Record,
+            ak._v2.highlevel.ArrayBuilder,
+        ),
+    ):
+        return array.type
+
+    elif isinstance(array, np.ndarray):
+        if len(array.shape) == 0:
+            return type(array.reshape((1,))[0])
+        else:
+            try:
+                out = type.dtype2primitive[array.dtype.type]
+            except KeyError:
+                raise TypeError(
+                    "numpy array type is unrecognized by awkward: %r" % array.dtype.type
+                )
+            out = ak._v2.types.PrimitiveType(out)
+            for x in array.shape[-1:0:-1]:
+                out = ak._v2.types.RegularType(out, x)
+            return ak._v2.types.ArrayType(out, array.shape[0])
+
+    elif isinstance(array, ak.layout.ArrayBuilder):
+        return NotImplementedError
+
+    elif isinstance(array, (ak._v2.contents.Content, ak._v2.record.Record)):
+        return array.form.type
+
+    elif isinstance(
+        array,
+        (
+            ak.highlevel.Array,
+            ak.highlevel.Record,
+            ak.highlevel.ArrayBuilder,
+            ak.layout.Content,
+            ak.layout.Record,
+        ),
+    ):
+        raise TypeError("do not use ak._v2.operations.convert.to_list on v1 arrays")
+
+    else:
+        raise TypeError("unrecognized array type: {0}".format(repr(array)))
+
+
+type.dtype2primitive = {
+    np.bool_: "bool",
+    np.int8: "int8",
+    np.int16: "int16",
+    np.int32: "int32",
+    np.int64: "int64",
+    np.uint8: "uint8",
+    np.uint16: "uint16",
+    np.uint32: "uint32",
+    np.uint64: "uint64",
+    np.float32: "float32",
+    np.float64: "float64",
+    np.complex64: "complex64",
+    np.complex128: "complex128",
+    np.datetime64: "datetime64",
+    np.timedelta64: "timedelta64",
+}
```

Comment:
The equivalent of this is in `ak._v2.types.numpytype._dtype_to_primitive_dict`, though that one takes dtypes to strings, rather than types to strings.

(It's necessary to use dtypes, rather than types, because a type like `np.datetime64` doesn't carry units, but a dtype like `np.dtype("datetime64[us]")` does.)